EIPS/eip-3978.md

```diff
@@ -0,0 +1,49 @@
+---
+eip: 3978
+title: Gas refunds on reverts
+description: Do not erase gas refunds on transaction subcall reverts, due users pay a lot of gas for storage non-modification.
+author: Anton Bukov (@k06a), Mikhail Melnik (@ZumZoom)
+discussions-to: https://ethereum-magicians.org/t/eip-3978-gas-refunds-on-reverts/7071/
+status: Draft
+type: Standards Track
+category: Core
+created: 2021-09-16
+---
+
+## Abstract
+
+Since [EIP-3298](https://eips.ethereum.org/EIPS/eip-3298) gas refunds works for storage restores only inside the same transaction. For example [ERC-20](./eip-20.md) `approve` + `transferFrom` flow between 2 smart contracts according to [EIP-2200](https://eips.ethereum.org/EIPS/eip-2200) and [EIP-2929](https://eips.ethereum.org/EIPS/eip-2929) will cost nearly to `21600` gas with gas refund counter `20000`. But in case of reverting this subcall (containing both `approve` and `transferForm`) gas refund will be erased, while smart contract storage will remain unmodified. I think it should keep storage access costs, but still refund modification costs.
+
+## Motivation
+
+Сurrent full cancelling of gas refunds on internal reverts is too unfair. Users pay for non-modification same cost as for modification.
+
+## Specification
+
+Let's count all SSTORE gas costs within every subcall, excluding access costs. And on reverting any subcall let's not erase refund counter, but:
```

Comment:
This needs to be much more clearly specified, and it needs to be phrased as an assertion not a suggestion (don't use terms like "let's" as it comes off as a suggestion rather than an assertion.  EIPs should be technical specifications that assert how things MUST behave.

Reply:
Thanks, this makes sense.